R/check.R

```diff
@@ -222,30 +224,227 @@ coerce_dt <- function(
+    if (length(select) > 0) { # if we have a select list ...
       select <- c(select, ".group") # ... add ".group" to it
     }
   }
 
   if (dates) {
-    dt[,               # cast-in-place to IDateTime (as.IDate)
+    dt[
+      , # cast-in-place to IDateTime (as.IDate)
       c("report_date", "reference_date") := .(
         as.IDate(report_date), as.IDate(reference_date)
       )
     ]
   }
 
-  if (length(select) > 0) {         # if selecting particular list ...
+  if (length(select) > 0) { # if selecting particular list ...
     return(dt[, .SD, .SDcols = c(select)][])
   } else {
     return(dt[])
   }
 }
 
+#' @title Check appropriateness of maximum delay
+#'
+#' @description Check if maximum delay specified by the user is long enough and
+#' raise potential warnings. This is achieved by computing the share of
+#' reference dates where the cumulative case count is below some aspired
+#' coverage.
+#'
+#' @details The coverage is with respect to the maximum observed case count for
+#' the corresponding reference date. As the maximum observed case count is
+#' likely smaller than the true overall case count for not yet fully observed
+#' reference dates (due to right truncation), only reference dates that are
+#' more than the maximum observed delay ago are included. Still, because we
+#' can only use the maximum observed delay, not the unknown true maximum
+#' delay, the computed coverage values should be interpreted with care, as they
+#' are only proxies for the true coverage.
+#'
+#' @inheritParams enw_filter_delay
+#' @inheritParams enw_preprocess_data
+#'
+#' @param data Output from [enw_preprocess_data()].
+#'
+#' @param cum_coverage The aspired percentage of cases that the maximum delay
+#' should cover. Defaults to 0.8 (80%).
+#'
+#' @param maxdelay_quantile_outlier Only reference dates sufficiently far in
+#' the past, determined based on the maximum observed delay, are included (see
+#' details). Instead of the overall maximum observed delay, a quantile of the
+#' maximum observed delay over all reference dates is used. This is more robust
+#' against outliers. Defaults to 0.97 (97%).
+#'
+#' @param warn Should a warning be issued if the cumulative case count is
+#' below `cum_coverage` for the majority of reference dates?
+#'
+#' @return A `data.table` with the share of reference dates where the
+#' cumulative case count is below `cum_coverage`, stratified by group.
+#'
+#' @family check
+#' @export
+#' @examples
+#' pobs <- enw_example(type = "preprocessed_observations")
+#' check_max_delay(pobs, max_delay = 20, cum_coverage = 0.8)
+check_max_delay <- function(data,
+                            max_delay = NULL,
+                            cum_coverage = 0.8,
+                            maxdelay_quantile_outlier = 0.97,
+                            warn = TRUE) {
+
+  if (is.null(max_delay)) {
+    max_delay <- data$max_delay
+  }
+  timestep <- data$timestep
+
+  if (warn == "internal") {
```

Comment:
Do we really need this complexity here vs just setting `warn = FALSE` when used internally? Generally I am a bit uncomfortable with arguments that change type (i.e here from logical to character)

Reply:
I feel like the case distinction is useful as we also want warnings when used internally, but slightly different warning messages compared to direct use of the function. I addressed the type change issue by adding a separate argument warn_internal that is FALSE by default and explaining it in the doc. Should be cleaner now.